### PR TITLE
Array persistence store with strict objects

### PR DESCRIPTION
**Pull request by mvorisek**
There are minor BC breaks like renamed array Action class, but these are normally not used by the end user thus the changes should be BC, also no changes in atk4/ui required or it was a bug previously.

Fixes several bugs like non strictly rectangular data, Array persistence throw if update is called on non existing ID...

### CODE REVIEW

File: tests/PersistentArrayTest.php [hunk 1]
```diff
@@ -392,13 +403,13 @@ public function testConditions()
...
             2 => ['name' => 'ABC8', 'code' => 12, 'country' => 'Ireland', 'active' => 0],
-            3 => ['code' => 13, 'country' => 'Latvia', 'active' => 1],
+            3 => ['name' => null, 'code' => 13, 'country' => 'Latvia', 'active' => 1],
             4 => ['name' => 'ABC6', 'code' => 14, 'country' => 'UK', 'active' => 0],
             5 => ['name' => 'ABC5', 'code' => 15, 'country' => 'UK', 'active' => 0],
             6 => ['name' => 'ABC4', 'code' => 16, 'country' => 'Ireland', 'active' => 1],
             7 => ['name' => 'ABC3', 'code' => 17, 'country' => 'Latvia', 'active' => 0],
             8 => ['name' => 'ABC2', 'code' => 18, 'country' => 'Russia', 'active' => 1],
-            9 => ['code' => 19, 'country' => 'Latvia', 'active' => 1],
+            9 => ['name' => null, 'code' => 19, 'country' => 'Latvia', 'active' => 1],
```
Comment (mvorisek) [inline tests/PersistentArrayTest.php:412]: this PR always make sure the data are rectangular, a positive side effect

File: src/Persistence/Array_/Db/Table.php [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    {
+        foreach ($newRowData as $columnName => $newValue) {
+            $this->assertValidValue($newValue);
+
+            // update index here
+        }
+    }
+
+    /**
+     * TODO rewrite with hash index support.
```
Comment (DarkSide666) [inline src/Persistence/Array_/Db/Table.php:182]: Yes this is very important for performance

File: tests/JoinArrayTest.php [hunk 3]
```diff
@@ -13,7 +13,18 @@ class JoinArrayTest extends AtkPhpunit\TestCase
 {
     private function getInternalPersistenceData(Persistence\Array_ $db): array
     {
-        return $this->getProtected($db, 'data');
+        $data = [];
+        /** @var Persistence\Array_\Db\Table $table */
+        foreach ($this->getProtected($db, 'data') as $table) {
+            foreach ($table->getRows() as $row) {
+                $rowData = $row->getData();
+                $id = $rowData['id'];
```
Comment (DarkSide666) [inline tests/JoinArrayTest.php:21]: hard-coded `'id'` field name here?

File: tests/StaticPersistenceTest.php [hunk 4]
```diff
@@ -113,9 +113,9 @@ public function testFieldTypes()
             'name' => 'hello',
             'test_int' => 123,
             'test_float' => 123.45,
-            'test_date' => new \DateTime(),
-            'test_array' => ['a', 'b', 'c'],
-            'test_object' => new \DateInterval('P1Y'),
+            // 'test_date' => new \DateTime(),
```
Comment (DarkSide666) [inline tests/StaticPersistenceTest.php:116]: Array persistence no more support non-scalar data types?

**Review by DarkSide666 — APPROVED** [review]
LGTM, but few questions

File: tests/JoinArrayTest.php (hunk 3 above)
Comment (mvorisek) [reply tests/JoinArrayTest.php:21]: yes - it is in specific test class, `id` here is safe to be assumed, the data are newly not indexed by some specific ID.

File: tests/StaticPersistenceTest.php (hunk 4 above)
Comment (mvorisek) [reply tests/StaticPersistenceTest.php:116]: It fully does, but you must typecast, static persistence does not typecast seed data.

File: src/Persistence/Array_/Db/Table.php (hunk 2 above)
Comment (mvorisek) [reply src/Persistence/Array_/Db/Table.php:182]: Will come in some next PR, previously, the array data were always iterated with the select action anyway.

**Pull request merged**